### PR TITLE
stat: Mann-Whitney U test and fractional ranks

### PR DESCRIPTION
**Pull request by kshedden**

### CODE REVIEW

**Comment by kortschak** [comment]
I think a "test" package would be worthwhile.

Also note that Austin Clements did a lot of work on Mann-Whitney U for benchstat. There's no reason we couldn't use that since we already have a license dependency on Go.

**Comment by kshedden** [comment]
I think this is the relevant code:

https://github.com/golang/perf/blob/6e6d33e29852650cab301f4dbeac9b7d67c6d542/internal/stats/utest.go

File: stat/stat.go [hunk 1]
```diff
@@ -1261,3 +1261,88 @@ func MeanVariance(x, weights []float64) (mean, variance float64) {
 	variance = (ss - compensation*compensation/sumWeights) / (sumWeights - 1)
 	return
 }
+
+// FractionalRank calculates the ranks of the elements in the input slice, using
```
[the remark this answers is not in the document]
Comment (kshedden) [reply stat/stat.go]: Ranks are estimates of quantiles at specific probability points, but using a different estimator (in the case of ties).  The gonum empirical Quantile function always returns one of the observed values, but fractional ranks are averages over several observed values (in the case of ties).  So there are connections between ranks and quantiles, but there enough differences that I think it makes more sense to implement them independently.  There are other use cases for fractional ranks, but the benchmark implementation doesn't separate the ranking from the U-test calculation.

[the remark this answers is not in the document]
Comment (btracey) [reply stat/stat.go]: The empirical estimator does, but there are other estimators. One of them is #605 which would have the behavior you suggest.

**Comment by kshedden** [comment]
I don't think the linearly interpolated quantiles can be easily used to get the fractional ranks.  But it's a moot point if the preference is to use the benchmark code instead of what we have here.

File: stat/stat.go [hunk 2]
```diff
@@ -1261,3 +1261,88 @@ func MeanVariance(x, weights []float64) (mean, variance float64) {
...
+			f += float64(j)
+		}
+		f /= float64(j - i)
+		for ; i < j; i++ {
+			ranks[inds[i]] = f
+		}
+	}
+}
+
+// MannWhitneyU conducts a two-sided Mann-Whitney U test for the null hypothesis
```
[the remark this answers is not in the document]
Comment (kortschak) [reply stat/stat.go]: I also prefer a "test" package.

**Comment by kortschak** [comment]
*ping*

**Comment by kshedden** [comment]
I'm not clear on what the next step is.  My understanding was that you want to close this PR without further action since the test is also in benchstat.

**Comment by vladimir-ch** [comment]
I'm closing this PR because it has gone stale. We're still interested in contributions, so please create a new PR if you still intend to submit this change.

**Pull request closed without merging**